setup.py

```diff
@@ -160,12 +160,18 @@ def main():
     # Raw paths relative to sourcetree root.
     files_outside_package_dir = {"cv2": ["LICENSE.txt", "LICENSE-3RD-PARTY.txt"]}
 
-    ci_cmake_generator = (
-        ["-G", "Visual Studio 14" + (" Win64" if is64 else "")]
-        if os.name == "nt"
-        else ["-G", "Unix Makefiles"]
-    )
-
+    if os.name == "nt": 
+        vs_env = os.environ.get("VISUAL_STUDIO")
+        if vs_env == "17":
+           generator_name = "Visual Studio 17 2022"
+           arch = "x64" if is64 else "Win32"
+           ci_cmake_generator = ["-G", generator_name, "-A", arch]
+        else:
+           generator_name = "Visual Studio 14" + (" Win64" if is64 else "")
+           ci_cmake_generator = ["-G", generator_name]
+    else:
+        ci_cmake_generator = ["-G", "Unix Makefiles"]
```

Comment:
It makes sense to override full generator name with environment variable and substitute it as is.